Hyprlinkr/DefaultRouteDispatcher.cs

```diff
@@ -108,5 +95,73 @@ public string RouteName
+                        
+            var controllerRouteAttribute = controllerType.GetCustomAttribute<RouteAttribute>(false);
+            if (controllerRouteAttribute != null)
+            {
+                return ExtractRoupleFromControllerRouteAttribute(callExpression, routeValues, controllerRouteAttribute);
+            }
+
+            var controllerName = controllerType.Name.ToLowerInvariant().Replace("controller", "");
+
+            var newRouteValues = new Dictionary<string, object>(routeValues);
+            newRouteValues["controller"] = controllerName;
+
+            return new Rouple(this.routeName, newRouteValues);
+        }
+
+        private Rouple ExtractRoupleFromControllerRouteAttribute(
+            MethodCallExpression callExpression, 
+            IDictionary<string, object> routeValues, 
+            RouteAttribute controllerRouteAttribute)
+        {
+            var controllerRouteName = this.routeName;
+            if (controllerRouteAttribute.Name != null)
+            {
+                controllerRouteName = controllerRouteAttribute.Name;
+            }
+
+            var newRouteValues = AddControllerNameAndMethodToRouteValues(callExpression, routeValues);
+            newRouteValues = RemoveControllerAndActionRouteValuesIfNeeded(controllerRouteAttribute, newRouteValues);
+            return new Rouple(controllerRouteName, newRouteValues);
+        }
+
+        private static Dictionary<string, object> RemoveControllerAndActionRouteValuesIfNeeded(
+            RouteAttribute controllerRouteAttribute, 
+            Dictionary<string, object> routeValues)
+        {
+            var controllerRouteTemplate = controllerRouteAttribute.Template;
+            if (controllerRouteTemplate != null)
+            {
+                if (!controllerRouteTemplate.Contains("{controller}"))
+                {
+                    routeValues.Remove("controller");
+                }
+                if (!controllerRouteTemplate.Contains("{action}"))
+                {
+                    routeValues.Remove("action");
+                }
+            }
+            return routeValues;
+        }
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Globalization", "CA1308:NormalizeStringsToUppercase", Justification = "This method should produce URIs with lower-case letters, so ultimately, it would have to invoke some sort of ToLower method.")]
+        private static Dictionary<string, object> AddControllerNameAndMethodToRouteValues(
+            MethodCallExpression callExpression, 
+            IDictionary<string, object> routeValues)
+        {
+            var newRouteValues = new Dictionary<string, object>(routeValues);
+            var controllerName = callExpression.Object.Type.Name.ToLowerInvariant().Replace("controller", "");
+            newRouteValues["controller"] = controllerName;
+            newRouteValues["action"] = callExpression.Method.Name.ToLowerInvariant();
```

Comment:
No problem, I should have added additional test cases.

Those methods manage the "controller" and "action" KVPs which may or may not be necessary to create the actual route Uri, depending on the actual Route Template.

The problem is that if they are put into the `RouteValues` dictionary and they are not needed, they will show up incorrectly as querystrings, which is not desirable.

I'll write the test cases that illustrate these issues.